File: solutions/cpp/secret-handshake/1/secret_handshake.cpp
#include "secret_handshake.h"
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

vector<string> secret_handshake::commands(int value) {
    vector<string> answer;
    static const vector<string> actions = {"wink", "double blink", "close your eyes", "jump"};

    for (int i = 0; i < 4; ++i) {
        if (value & (1 << i)) {
            answer.push_back(actions[i]);
        }
    }

    if (value & (1 << 4)) {
        reverse(answer.begin(), answer.end());
    }

    return answer;
}

File: solutions/haskell/secret-handshake/1/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits (testBit)

data Action = Wink
            | DoubleBlink
            | CloseYourEyes
            | Jump
            | Reverse
            deriving (Show, Eq, Enum, Bounded)

handshake :: Int -> [String]
handshake n = applyReverseIfNeeded (map actionToString actions)
  where
    actions = [action | action <- [Wink .. Jump], n `testBit` fromEnum action]

    applyReverseIfNeeded :: [String] -> [String]
    applyReverseIfNeeded actionsList
      | n `testBit` fromEnum Reverse = reverse actionsList
      | otherwise                    = actionsList

    actionToString :: Action -> String
    actionToString Wink          = "wink"
    actionToString DoubleBlink   = "double blink"
    actionToString CloseYourEyes = "close your eyes"
    actionToString Jump          = "jump"
    actionToString Reverse       = undefined  -- Reverse should never be converted to a string
